### DISCUSSION TITLE
Working without server

### DISCUSSION THREAD

**Discussion by Cattn** [EmulatorJS/EmulatorJS#429] [Q&A] [answered]
So I know the official page has a way to package a rom as a data string (I think that's what it does) and then load it in without a web server, I was wondering how you were able to convert the rom to the long data, and if you could point out either some docs about how to use the window.gameData, or something else of the like. I looked through docs and didn't see much about it

**Comment by Cattn** [comment, marked as answer]
Any help would be appreciated!

**Reply by ethanaobrien** [reply]
Sorry, so all it is is the way you set the game url. It works offline because it sets the path to data to https://rawcdn.githack.com/EmulatorJS/EmulatorJS/main/data/, so we can make requests toward non javascript files. Second, `window.gameData` is just a `Uint8Array` that, on runtime, is converted to a blob. The exact code is
```
URL.createObjectURL(new Blob([Uint8Array.from(window.gameData)]));
```

**Reply by Cattn** [reply]
Thanks so much for the explanation! I'm trying to get the emulator to work locally with the game rom separate, will that work or will it have to be converted into unit8array

**Reply by ethanaobrien** [reply]
Nope, there is no way we can fetch the rom that way because of the browser security policy. It must be converted to a Uint8Array. Sorry

**Reply by Cattn** [reply]
Basically the goal is that the user can upload a rom to a folder and then it can load locally (whether it's converted into a Unit8Array or another way) I can get down grabbing files uploaded locally to the folder, and names just not sure how to convert it properly in this case

**Reply by Cattn** [reply]
> Nope, there is no way we can fetch the rom that way because of the browser security policy. It must be converted to a Uint8Array. Sorry

I figured, makes sense

**Reply by Cattn** [reply]
Could you show what you did to convert the rom to a Unit8Array, as it's done in the website when it generates an html file?

**Reply by ethanaobrien** [reply]
```
JSON.stringify(Array.from(new Uint8Array(await (new Blob([file])).arrayBuffer())));
```

**Reply by Cattn** [reply]
thank you!

**Reply by Cattn** [reply]
Whenever I upload a file it only outputs in a length of 17 in the array
This is what I'm trying currently, It is able to get the file fine, I'm not sure if I'm doing something wrong past that.
`let input = document.querySelector('input')`
`input.addEventListener('change', () => {`
`let file = input.files;`
`async function convertFile() {`
`let test = JSON.stringify(Array.from(new Uint8Array(await (new Blob([file])).arrayBuffer())));`
`console.log(test);`
`console.log(file);`
`}`
`convertFile();`
`});`

**Reply by ethanaobrien** [reply]
let file = input.files; needs to be let file = input.files[0];

**Reply by Cattn** [reply]
Thank you so much that solved it!
